cli/graph.rs

```diff
@@ -0,0 +1,908 @@
+            } else {
+              let dep = self.dependencies.entry(import).or_default();
+              dep.maybe_type = Some(specifier);
+            }
+          }
+        }
+      }
+    }
+
+    // Parse out all the syntactical dependencies for a module
+    let dependencies = parsed_module.analyze_dependencies();
+    for desc in dependencies.iter() {
+      let location = Location {
+        filename: self.specifier.to_string(),
+        col: desc.col,
+        line: desc.line,
+      };
+      let specifier =
+        self.resolve_import(&desc.specifier, Some(location.clone()))?;
+
+      // Parse out any `@deno-types` pragmas and modify dependency
+      let maybe_types_specifier = if !desc.leading_comments.is_empty() {
+        let comment = desc.leading_comments.last().unwrap();
+        if let Some(deno_types) = parse_deno_types(&comment.text).as_ref() {
+          Some(self.resolve_import(deno_types, Some(location))?)
+        } else {
+          None
+        }
+      } else {
+        None
+      };
+
+      let dep = self
+        .dependencies
+        .entry(desc.specifier.to_string())
+        .or_default();
+      if desc.kind == DependencyKind::ExportType
+        || desc.kind == DependencyKind::ImportType
+      {
+        dep.maybe_type = Some(specifier);
+      } else {
+        dep.maybe_code = Some(specifier);
+      }
+      if let Some(types_specifier) = maybe_types_specifier {
+        dep.maybe_type = Some(types_specifier);
+      }
+    }
+
+    self.maybe_parsed_module = Some(parsed_module);
+    Ok(())
+  }
+
+  fn resolve_import(
+    &self,
+    specifier: &str,
+    maybe_location: Option<Location>,
+  ) -> Result<ModuleSpecifier> {
+    let maybe_resolve = if let Some(import_map) = self.maybe_import_map.clone()
+    {
+      import_map
+        .borrow()
+        .resolve(specifier, self.specifier.as_str())?
+    } else {
+      None
+    };
+    let specifier = if let Some(module_specifier) = maybe_resolve {
+      module_specifier
+    } else {
+      ModuleSpecifier::resolve_import(specifier, self.specifier.as_str())?
+    };
+
+    let referrer_scheme = self.specifier.as_url().scheme();
+    let specifier_scheme = specifier.as_url().scheme();
+    let location = maybe_location.unwrap_or(Location {
+      filename: self.specifier.to_string(),
+      line: 0,
+      col: 0,
+    });
+
+    // Disallow downgrades from HTTPS to HTTP
+    if referrer_scheme == "https" && specifier_scheme == "http" {
+      return Err(InvalidDowngrade(specifier.clone(), location).into());
+    }
+
+    // Disallow a remote URL from trying to import a local URL
+    if (referrer_scheme == "https" || referrer_scheme == "http")
+      && !(specifier_scheme == "https" || specifier_scheme == "http")
+    {
+      return Err(InvalidLocalImport(specifier.clone(), location).into());
+    }
+
+    Ok(specifier)
+  }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct Stats(Vec<(String, u128)>);
```

Comment:
Use type alias instead? That way there will be no need for custom `Deserialize` implementation

Reply:
I've got a custom `fmt:Display` as well.  It is nice to have it all be self contained, IMO.